pkg/ebpf/printk_patcher.go

```diff
@@ -0,0 +1,168 @@
+
+			// We found the call to bpf_trace_printk, now we need to find
+			// the string on the stack and patch it.
+			// For that, find first the second register, which is the second argument
+			// to the call, which is the length of the string.
+			var lengthLoadIns *asm.Instruction
+			for i := idx - 1; i >= maxLookback; i-- {
+				candidate := &p.Instructions[i]
+				if candidate.OpCode == movImmOpCode && candidate.Dst == asm.R2 {
+					lengthLoadIns = candidate
+					break
+				}
+			}
+			if lengthLoadIns == nil {
+				err = log.Warnf("Could not find length load instruction for bpf_trace_printk call %d in %s", idx, p.Name)
+				continue // Skip this call instruction
+			}
+
+			// Now we have to find in which part the stack is the string being stored
+			// For that we need to find the mov instruction that puts the stack pointer
+			// into r1 and then the add that modifies the stack offset
+			var stackOffsetIns *asm.Instruction
+			for i := idx - 1; i >= maxLookback && stackOffsetIns == nil; i-- {
+				candidate := &p.Instructions[i]
+				if candidate.OpCode == movRegOpCode && candidate.Dst == asm.R1 && candidate.Src == asm.RFP {
+					// Ok, so we found the instruction that loads the stack pointer into r1
+					// From that, advance until we find the add instruction that modifies the stack offset
+					for j := i + 1; j < idx; j++ {
+						candidate = &p.Instructions[j]
+						if candidate.OpCode == addImmOpCode && candidate.Dst == asm.R1 {
+							stackOffsetIns = candidate
+							break
+						}
+					}
+				}
+			}
+			if stackOffsetIns == nil {
+				err = log.Warnf("Could not find stack offset instruction for bpf_trace_printk call %d in %s", idx, p.Name)
+				continue
+			}
+			newlineOffset := int16(stackOffsetIns.Constant + lengthLoadIns.Constant - 1) // -1 because the last character is the null character
+
+			// Now find which store instruction is responsible for putting the newline character
+			stringStoreInsIdx := -1
+			inInstructionOffset := 0
+			for i := idx - 1; i >= maxLookback; i-- {
+				candidate := &p.Instructions[i]
+				if candidate.OpCode.Class() == asm.StXClass && candidate.Src == asm.R1 && candidate.Dst == asm.RFP {
+					if candidate.OpCode.Size() == asm.InvalidSize {
+						err = log.Warnf("BUG: store instruction %v returned asm.InvalidSize", candidate)
+						continue
+					}
+
+					minOffset := candidate.Offset
+					maxOffset := minOffset + int16(candidate.OpCode.Size().Sizeof())
+
+					if newlineOffset >= minOffset && newlineOffset <= maxOffset {
+						// We found the store instruction that loads the newline character, exit the loop
+						stringStoreInsIdx = i
+						inInstructionOffset = int(newlineOffset - minOffset)
+						break
+					}
+				}
+			}
+			if stringStoreInsIdx == -1 {
+				err = log.Warnf("Could not find store instruction for bpf_trace_printk call %d in %s", idx, p.Name)
+				continue
+			}
+
+			// Now try to find the load instruction that loads the string into the R1 register
```

Comment:
Could you add example ebpf instructions for each of these blocks? That might help folks understand what a full example `log_debug` calls looks like, and what we are searching for in each step.